redis/subConn.go

```diff
@@ -1,135 +1,192 @@
+			MaxActive:      3,
+			MaxIdle:        1,
+			SetReadTimeout: false,
+		}),
+		outputChan:      make(chan redis.Message, channelsBuffersSize),
+		subscribeChan:   make(chan []interface{}, channelsBuffersSize),
+		unsubscribeChan: make(chan []interface{}, channelsBuffersSize),
 	}
 
-	if err := subConn.resetConn(); err != nil {
+	err := startMainLoop(subConn)
+	if err != nil {
 		return nil, err
 	}
-	go subConn.mainLoop()
-
 	return subConn, nil
 }
 
 func (c *subConn) PSubscribe(patterns []interface{}) error {
-	if err := c.currConn.PSubscribe(patterns...); err != nil {
-		return errors.WithStack(err)
+	if len(patterns) == 0 {
+		return errors.New("Must send at least one pattern to subscribe")
 	}
 
-	c.mu.Lock()
-	defer c.mu.Unlock()
-
-	for _, pattern := range patterns {
-		c.subscribedPatterns[pattern] = true
+	select {
+	case c.subscribeChan <- patterns:
+		return nil
+	case <-time.After(subscribeTimeout):
+		return errors.New("Timeout sending patterns to subscriptions channel")
 	}
-	return nil
 }
 
 func (c *subConn) PUnsubscribe(patterns []interface{}) error {
 	if len(patterns) == 0 {
 		return errors.New("Must send at least one pattern to unsubscribe")
 	}
 
-	if err := c.currConn.PUnsubscribe(patterns...); err != nil {
-		return errors.WithStack(err)
-	}
-
-	c.mu.Lock()
-	defer c.mu.Unlock()
-
-	for _, pattern := range patterns {
-		delete(c.subscribedPatterns, pattern)
-	}
+	// Since this is almost a clean-up done after the consumers are done with us,
+	// we don't have a timeout here so we don't risk leaving an inconsistent state.
+	// If we have any problems with this, consider moving this to a background
+	// routine so we don't block the caller.
+	c.unsubscribeChan <- patterns
 	return nil
 }
 
 func (c *subConn) ReceiveChan() <-chan redis.Message {
 	return c.outputChan
 }
 
-func (c *subConn) mainLoop() {
-	var (
-		pingTicker      = time.NewTicker(pingDelay)
-		pongTimeoutChan <-chan time.Time
+type pubSubLoopState struct {
+	parent             *subConn
+	subscribedPatterns map[interface{}]bool
+
+	pingTicker      <-chan time.Time
+	pongTimeoutChan <-chan time.Time
 
-		done    = make(chan struct{})
-		msgChan = connReceiveChan(c.currConn, done)
-	)
-	recover := func() {
-		close(done)
+	currConn          *redis.PubSubConn
+	msgChan           <-chan interface{}
+	cancelCurrMsgChan func()
+}
 
-		c.recoverConn()
-		done = make(chan struct{})
-		msgChan = connReceiveChan(c.currConn, done)
+func startMainLoop(parent *subConn) error {
+	loopState := &pubSubLoopState{
+		parent:             parent,
+		subscribedPatterns: map[interface{}]bool{},
+		pingTicker:         time.Tick(pingDelay),
 	}
-	for {
-		select {
-		case <-pingTicker.C:
-			err := c.currConn.Ping("")
-			if err == nil {
-				pongTimeoutChan = time.Tick(pongTimeout)
-			} else {
-				logError(err, "pubsub_ping_error", "", "", "Redis error pinging pub/sub connection")
-				recover()
+	if err := loopState.resetConn(); err != nil {
+		return err
+	}
+
+	go func() {
```

Comment:
I think we already discussed this, but is it ok to never stop the connection loop?

Reply:
In our current use-cases, the Redis (pub-sub) instances are created as singleton clients that are never stopped, so we don't really need a stopping/cancellation logic here and I opted for avoiding the boilerplate. If you think it'd make sense having it though, I could do it, as it's mostly a soft preference of mine.